### PR TITLE
Suggestion : Randomized related challenges view

### PR DESCRIPTION
**Pull request by lobertrand**
Hi! The "Try a Challenge!" section on a challenge page only shows the first 2 challenges from the "relatedChallenges" array. So putting more than 2 challenges in this array seems kind of odd. Unless we randomize the challenges that are shown on the page.

So this is my suggestion :smile:  I don't know if that's a good idea. One drawback of this is that this doesn't seem to work great with SSG (Static Site Generation).

Preview example : https://deploy-preview-572--codingtrain.netlify.app/challenges/162-self-avoiding-walk

### CODE REVIEW

**Comment by shiffman** [comment]
Hi @lobertrand! This is so great! It's actually something I've been thinking about. Even though the passenger showcase scrolls it might be nice to have that randomized too. There are probably other places on the site I could imagine featuring something at random (or maybe random from a list of options, like the "featured challenges" or "featured showcase" on the landing page.)

I think the question here is whether it should be randomized when the site builds (so everything is kept completely static) or in realtime when the page is requested. I would prefer the latter but I'm not sure if the former is more appropriate for a gatsby site.

@runemadsen do you think you might be able to point us in the right direction here?

**Comment by lobertrand** [comment]
I'm facing a weird behavior when suffling the array of challenges : the array of challenges seems to be shuffled on each page refresh because the title, description and dates are changing, but some parts stay static (thumbnails and links).

Looking further, it seems like **text nodes** are shuffled properly, but not **attribute values** : 
frankenstein_dynamic_content
In particular, we can see that the date text node (jan 16, 2017) matches the A* challenge, but the value of the `datetime` attribute never changes (just like the thumbnail) and matches the maze generation challenge.

This behavior only occurs when running `npm run build`, not in dev mode.

**Comment by runemadsen** [comment]
Excellent! Thanks for doing this. I'm all for choosing a random challenge on the client side if we can stay under a length of `30` challenges. Since we need to load all the challenges on the client and only show one random one, it's adding `kb` to the page, but as long as we don't go wild, I think it's perfectly fine.

@lobertrand The reason why your code is not working is that you're using `Math.random()`. So when Gatsby renders the React components on the server side, it generates on random number, and when it hydrates on the client, it generates another random number.  This leads to hydration errors because React don't have matching states for the server and the client.

This is a general problem with SSR and there are multiple ways to solve it:

1. **Only render the related challenges on the client**. This is done by something like this:

```
const [relatedChallenges, setRelatedChallenges] = useState(null);

useEffect(() => {
  setRelatedChallenges([something, something]);
}, [challenges]);

return <div>
  {relatedChallenges]
</div>
```

The downside to this is that Google won't be able to index the related challenges since they are not in the initial HTML.

2. **Make a deterministic random**. This is the best solution, but it's often hard to pull off since you want a deterministic random that returns the same random numbers on the server and on the client, even when the site builds on one day on the server and is rendered on the client another day. 

Perhaps I would go with 1. for now, but you could also try 2. as a coding challenge 😂

**Comment by lobertrand** [comment]
@runemadsen Thank you for your insights! I'm gonna try to get to an optimal solution :) I found this article but haven't tried it yet : https://blog.logrocket.com/fixing-gatsbys-rehydration-issue/

My work on this project is my first concrete experience with React, Gatsby, SSG/SSR and GraphQL, so I'm learning a lot from this 😁

**Comment by shiffman** [comment]
Thank you @runemadsen for your super helpful and detailed reply!

@lobertrand if you need any additional support or feel you would like to hand this off to someone else, feel free to let me know anytime!

**Comment by runemadsen** [comment]
Great! Let me know how it goes. I'm happy to give feedback if needed :)

**Comment by lobertrand** [comment]
@shiffman @runemadsen Hi :) Here is my progress on this topic :

### 1. Rendering related challenges only on the client

I tested your first solution and it works :) The downsides of it are - like you said - that it's not ideal for SEO and that users could see this (if they scrolled fast enough or with a slow connection) :

Section appears blank and then challenges appear

### 2. Rendering related challenges both on the server and on the client

One alternative would be to render the list in some random order on SSR and in some other random order on hydration. This should be better for SEO, but users could see this :

A first shuffle appears and then is replaced with another one on each page refresh

For this to work without the problem I mentioned before, I added a ["key" attribute to force a fresh rerender](https://medium.com/@albertogasparin/forcing-state-reset-on-a-react-component-by-using-the-key-prop-14b36cd7448e) of this portion of the tree during hydration rather than letting React badly reconcile static and dynamic content together.

```jsx
const isSSR = useIsSSR();
...
<div className={css.challenges} key={isSSR}>
```

### 3. Deterministic random generator

I guess the deterministic random number generator you proposed would have to be seeded by a **different seed on each build** (and this same seed would have to be used during hydration). Otherwise, we would get the exact same shuffle on each build.

If this is feasable, this would have the advantage to not cause any flickering/loading. The only downside would be that the order would only change when a new build occurs.

By the way :
- I added the [`DEV_SSR`](https://github.com/gatsbyjs/gatsby/discussions/28138) flag to my local `gatsby-config.js` file to enable SSR in dev mode. This really helps to test this out :)
- For these screen captures, I activated the "Fast 3G" mode of Chrome DevTools to simulate a slow connection.

**Comment by shiffman** [comment]
Wow! This is so cool, thank you @lobertrand! I would defer to @runemadsen in terms of which solution is optimal and best. Two questions / thoughts:

1. Is there one that feels right to you in terms of maintenance of the code?
2. Is there one that would be good in terms of setting a precedent for other similar "randomizers" on the website. For example, a "featured showcase" randomizer on the homepage or "showcase order" randomizer on challenge pages?

I'm not too concerned about SEO (tell me if I should be!) given that these are "related challenges" and not the core content of the page.

**Comment by lobertrand** [comment]
@shiffman Glad to help :)

> 1. Is there one that feels right to you in terms of maintenance of the code?

The first solution seems to me like the most straightforward way to implement. The second one requires a little trick (`isSSR` boolean) but is still understandable with some comments. I think the third one would really require some hacks and complexify the understanding of the code.

> 2. Is there one that would be good in terms of setting a precedent for other similar "randomizers" on the website. For example, a "featured showcase" randomizer on the homepage or "showcase order" randomizer on challenge pages?

I guess the first one would be a great fit for secondary content that's "below the fold" (we only see it when we scroll the page so it has time to load). So it would fit the "Passenger showcase highlight" on the homepage as well as the "Passenger showcase" on the challenge page.

> I'm not too concerned about SEO (tell me if I should be!) given that these are "related challenges" and not the core content of the page.

That's what I was thinking too :)

**Comment by runemadsen** [comment]
Amazing! this is great progress. I would also recommend going with the first and simplest implementation. The solution to the flicker is to render an empty div with the same height as the challenges panel when the challenges are `null`. Since the challenges panel is a set height, this will be fairly simple to do. You can even do this in cases where you don't know the height of the element you're randomizing. You'll see this on [our website](https://designsystems.international/) where we have placeholder elements before the projects are selected. There's a bit of a flicker, but it's much less than going from nothing to filled in content.

**Comment by shiffman** [comment]
Wonderful! @lobertrand let me know if you need any additional support! I am so grateful for this contribution!

**Comment by lobertrand** [comment]
@shiffman @runemadsen I used "empty" challenges as placeholders during the load of client side code, which gives this effect (on a slow connection) : 

skeleton_load_two

and if there is just one challenge in the "relatedChallenges" array : 

skeleton_load_one

Let me know if there is anything I can enhance :smiley: 

By the way, this component is used on two pages : the challenge page and the track video page (if related challenges exist for the given video). I applied the shuffle on both pages for simplicity, it that okay?

**Comment by shiffman** [comment]
Wonderful, is this ready to review? I really like how it works. The idea is to have related challenges in all of the videos but b/c this wasn't data on the original site it wasn't prioritized during the porting (#262). Maybe I can bring this up on [tomorrow's live stream](https://www.youtube.com/watch?v=EiYhMpZ_43M).

We could even consider picking a random challenge to suggest if none are specified but maybe that's not in the spirit of this feature.

**Comment by shiffman** [comment]
Ah I see your "ready to review"! I've tested it and it's good to go. @runemadsen any additional comments or thoughts before I merge?

File: src/utils/index.js
Comment (runemadsen) [inline src/utils/index.js:82]: I think it would make more sense to call this hook something like `isFirstRender`, since it will render true for the first render on the server and for the first render on the client too. So something like the `isClient` boolean will confuse people because it will be `false` on the first render in the client. I propose renaming the hook to `isFirstRender` and then just returning a single boolean.

File: src/components/ChallengesPanel.js
Comment (runemadsen) [inline src/components/ChallengesPanel.js:86]: I originally thought this was a little weird, but realized that you do this to make sure there's the same number of empty challenges as there are challenges. So I would make that clear in the comment, so someone doesn't replace this with a simple `[{}, {}]`

**Review by runemadsen — APPROVED** [review]
I added a few small comments, but other than that it's beautiful and ready to go!

File: src/utils/index.js
Comment (lobertrand) [reply src/utils/index.js]: You're absolutely right! I myself got confused with how many times this value was true ^^ It's much clearer now. I renamed it to `useIsFirstRender`, starting with `use` to make it clear that this is a custom hook. Since this function encapsulates a hook, [rules of hooks](https://reactjs.org/docs/hooks-rules.html) apply to it (see [this page](https://reactjs.org/docs/hooks-custom.html) for more info).

File: src/components/ChallengesPanel.js
Comment (lobertrand) [reply src/components/ChallengesPanel.js]: Thanks a lot for the feedback, I added a more precise comment :)

**Comment by lobertrand** [comment]
Once this is merged and if it looks good, I'd be happy to work on the implementation of the other "randomizers" we talked about :smiley:

**Comment by shiffman** [comment]
Ok! Here we go @lobertrand! And yes, I would be thrilled to add more randomizers throughout the site! Feel free to DM me or open an issue to discuss further!

**Pull request merged**
